src/wast-parser.cc

```diff
@@ -0,0 +1,2013 @@
+    case TokenType::AssertInvalid:
+    case TokenType::AssertMalformed:
+    case TokenType::AssertReturn:
+    case TokenType::AssertReturnArithmeticNan:
+    case TokenType::AssertReturnCanonicalNan:
+    case TokenType::AssertTrap:
+    case TokenType::AssertUnlinkable:
+    case TokenType::Get:
+    case TokenType::Invoke:
+    case TokenType::Module:
+    case TokenType::Register:
+      return true;
+    default:
+      return false;
+  }
+}
+
+bool IsEmptySignature(const FuncSignature* sig) {
+  return sig->result_types.empty() && sig->param_types.empty();
+}
+
+void ResolveFuncTypes(Module* module) {
+  for (ModuleField& field : module->fields) {
+    Func* func = nullptr;
+    if (field.type == ModuleFieldType::Func) {
+      func = dyn_cast<FuncModuleField>(&field)->func;
+    } else if (field.type == ModuleFieldType::Import) {
+      Import* import = dyn_cast<ImportModuleField>(&field)->import;
+      if (import->kind == ExternalKind::Func) {
+        func = import->func;
+      } else {
+        continue;
+      }
+    } else {
+      continue;
+    }
+
+    // Resolve func type variables where the signature was not specified
+    // explicitly, e.g.: (func (type 1) ...)
+    if (func->decl.has_func_type && IsEmptySignature(&func->decl.sig)) {
+      FuncType* func_type = module->GetFuncType(func->decl.type_var);
+      if (func_type) {
+        func->decl.sig = func_type->sig;
+      }
+    }
+
+    // Resolve implicitly defined function types, e.g.: (func (param i32) ...)
+    if (!func->decl.has_func_type) {
+      Index func_type_index = module->GetFuncTypeIndex(func->decl.sig);
+      if (func_type_index == kInvalidIndex) {
+        auto func_type = new FuncType();
+        func_type->sig = func->decl.sig;
+        module->AppendField(new FuncTypeModuleField(func_type, field.loc));
+      }
+    }
+  }
+}
+
+void AppendInlineExportFields(Module* module,
+                              ModuleFieldList* fields,
+                              Index index) {
+  Location last_field_loc = module->fields.back().loc;
+
+  for (const ModuleField& field: *fields) {
+    auto* export_field = cast<ExportModuleField>(&field);
+    export_field->export_->var = Var(index, last_field_loc);
+  }
+
+  module->AppendFields(fields);
+}
+
+}  // End of anonymous namespace
+
+WastParser::WastParser(WastLexer* lexer,
+                       ErrorHandler* error_handler,
+                       WastParseOptions* options)
+    : lexer_(lexer), error_handler_(error_handler), options_(options) {}
+
+void WastParser::Error(Location loc, const char* format, ...) {
+  errors_++;
+  va_list args;
+  va_start(args, format);
+  WastFormatError(error_handler_, &loc, lexer_, format, args);
+  va_end(args);
+}
+
+Token WastParser::GetToken() {
+  if (tokens_.empty())
+    tokens_.push_back(lexer_->GetToken(this));
+  return tokens_.front();
+}
+
+Location WastParser::GetLocation() {
+  return GetToken().loc;
+}
+
+TokenType WastParser::Peek() {
```

Comment:
I thought about this, but since we only need two tokens of lookahead, I thought it would be better to make that clear by not making it general. Guess I should add some more documentation about all this.